// generator-jhipster-material/generators/app/index.js
'use strict';

var generators = require('yeoman-generator');
var jhipster = require('generator-jhipster');
var packagejs = require(__dirname + '/../../package.json');
var _ = require('underscore.string');
var chalk = require('chalk');
var path = require('path');

// Stores JHipster variables
var jhipsterVar = {moduleName: 'react'};

// Stores JHipster functions
var jhipsterFunc = {};

/* Constants use throughout */
const constants = require('../generator-constants'),
    QUESTIONS = constants.QUESTIONS;

var currentQuestion = 0;
var configOptions = {};

module.exports = generators.Base.extend({
  initializing: {

    displayLogo: function () {
      this.log(chalk.white('\nWelcome to ' + chalk.bold('JHipster Material') + '! ' + chalk.yellow('v' + packagejs.version + '\n')));
    },

    composeModule: function (args) {
      /* compose with Jhipster module to get access to reusable functions from JHipster */
      this.composeWith('jhipster:modules', {
        options: {
          jhipsterVar: jhipsterVar,
          jhipsterFunc: jhipsterFunc,
          skipValidation: true
        }
      });
    },

    setupClientVars : function () {
      this.useSass = this.config.get('useSass');
      this.appName = this.config.get('appName');
      this.frontendBuilder = this.config.get('frontendBuilder');
      this.enableTranslation = this.config.get('enableTranslation');
      if(this.useSass !=null && this.frontendBuilder != null && this.enableTranslation != null) {
        this.existingProject = true;
      }
    },

  },

  prompting: {
    askForModuleName: function () {
            if (this.existingProject) {
                return;
            }
            this.askModuleName(this, ++currentQuestion, QUESTIONS);
            configOptions.lastQuestion = currentQuestion;
        }
    }
  },

  configuring: {

    composeServerApp: function (args) {
      this.log(chalk.green.bold('\n-----------------------* Server side options *-----------------------\n'));
      /* Compose server side application using JHipster */
      this.composeWith('jhipster:app', {
        options: {
          'client-hook': !this.skipClient,
          configOptions: configOptions
        }
      });
    },

    composeClientApp: function (args) {
      this.log(chalk.green.bold('\n-----------------------* Client side options *-----------------------\n'));
      /* Compose client side application using client sub generator */
      this.composeWith('jhipster-material:client', {
          options: {
            'skip-install': this.options['skip-install'],
            configOptions: configOptions
          }
      }, {
          local: require.resolve('../client')
      });
    },

  },

  configuring : {
    saveConfig: function () {
      this.config.set('appName', this.appName);
      this.config.set('useSass', this.useSass);
      this.config.set('frontendBuilder', DEF_CLIENT_BUILD);
      this.config.set('enableTranslation', this.enableTranslation);
    },

    setupVars: function () {
      this.baseName = this.appName;
      this.slugifiedBaseName = _.slugify(this.baseName);
    }
  },

  writing: {
    /* write client side files */
    writeCommonFiles : function () {
      console.log(this.slugifiedBaseName)
      this.template('_package.json','package.json');
      this.copy('.babelrc','.babelrc');
      this.copy('.gitignore','.gitignore');
      this.copy('.eslintrc', '.eslintrc');
    },

    writeWebpackFiles : function() {
      this.copy('_webpack-production.config.js', 'webpack-production.config.js');
      this.copy('_webpack-dev-server.config.js', 'webpack-dev-server.config.js');
      this.copy('_server.js', 'server.js')
    },

    writeMainFiles : function () {
      this.template(WEB_SRC + '_index.html',WEB_SRC + 'index.html');
      this.copy(WEB_SRC + '404.html',WEB_SRC + '404.html');
      this.copy(WEB_SRC + 'favicon.ico',WEB_SRC + 'favicon.ico');
      this.copy(WEB_SRC + 'htaccess.txt',WEB_SRC + '.htaccess');
      this.copy(WEB_SRC + 'robots.txt',WEB_SRC + 'robots.txt');
      this.copy(WEB_SRC + 'content/css/_main.css',WEB_SRC + 'content/css/main.css');
    },

    writeAppFiles : function () {
      this.copy(WEB_SRC + 'app/_app.jsx',WEB_SRC + 'app/app.jsx');
      this.template(WEB_SRC + 'app/_Main.jsx',WEB_SRC + 'app/Main.jsx');
    }
  },

  install: function () {
    if(this.options['skip-install']){
      this.log(chalk.green.bold('\nRun command `npm install` on the project root to install required dependencies\n'));
    } else {
      this.log(chalk.green.bold('\nRunning NPM install. If this fails run command `npm install` on the project root\n'));
      this.npmInstall();
    }
  }
});
